### ISSUE TITLE
don't print additive part of matrix multiplication if it's zero

### ISSUE THREAD

**Issue by ghorn** [casadi/casadi#1311]
```python
X = MX.sym('X', 2, 2)
print mul(X, X)
```
emits
```
(zeros(2x2, dense)+mul(X, X))
```
I vote for dropping the zeros part (including the parenthesis)

**Comment by jaeandersson** [comment]
Against. The multiplication node calculates Z + mul(X, Y) so the above is what is actually being calculated. A bit more compact syntax could be `(zeros(2, 2)+mul(X, X))`

**Comment by jgillis** [comment]
I vote in support of Greg

**Comment by ghorn** [comment]
> The multiplication node calculates Z + mul(X, Y) so the above is what is actually being calculated.

So you don't actually simplify out the case when Z==0, right? (I think this is how BLAS works too btw)

**Comment by jaeandersson** [comment]
Right, no simplification. Note that a matrix multiplication is calculating sums of products which means that to calculate `mul(X, Y)`, you'd first need to set the partial sums to zero. Also, `f(X, Y, Z) = Z + mul(X, Y)` has a more natural AD rules. By design, the result has the same sparsity as Z so with live variables (this is the normal case), the operation actually calculates: `Z += mul(X, Y)`.

**Comment by jaeandersson** [comment]
I think the MX printout should reflect what is actually being calculated. I think the printout is natural since the calculation of the above will in general be:
```
Z = zeros(2,2)
Z += mul(X, X)
```

**Comment by jaeandersson** [comment]
FWIW, BLAS DGEMM calculates `C := alpha*mul(A,B) + beta*C` with optional transpose of A and B. So it corresponds to CasADi's definition with alpha=beta=1.
